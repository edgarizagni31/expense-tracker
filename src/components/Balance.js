import React from "react";
import PropTypes from 'prop-types';
import { getStatistics } from "../helpers/getStatistics";

export const Balance = ({ transactions }) => {
  const { totalIcome, totalExpense, balance } = getStatistics( transactions );

  return (
    <section className = 'mb-4 container '>
      <div className = 'mb-2'>
        <p className = 'mb-0 p-0'> Tu balance </p>
        <p className = 'fw-bold fs-4 p-0'>S/ {balance}</p>
      </div>
      <article className = 'row'>
        <div className = 'col shadow-sm text-center bg-white me-1 '>
          <p className = 'mt-1'>Ingresos</p>
          <p className = 'text-success'>S/ {totalIcome}</p>
        </div>
        <div className = 'col shadow-sm text-center bg-white'>
          <p className = 'mt-1'>Egresos</p>
          <p className = 'text-danger'>S/ { Math.abs(totalExpense) }</p>
        </div>
      </article>
    </section>
  );
};

Balance.propTypes = {
  transactions: PropTypes.array.isRequired
}