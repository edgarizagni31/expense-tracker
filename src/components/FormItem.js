import React from 'react'

export const FormItem = ( { label, id, placeholder, handleInputChange, formText } ) => {
    return (
        <div className = 'mb-3' id = 'formItem'>
            <label htmlFor = {id} className = 'form-label mb-0'>{label}</label>

            { formText !== "" && <p className = 'form-text mt-0'> { formText } </p> }

            <input 
              type = 'text'
              id = { id }
              name = { id }
              placeholder = { placeholder } 
              className = 'form-control'
              onChange = { handleInputChange }
              autoComplete = 'off'
            />
        </div>
    )
}
