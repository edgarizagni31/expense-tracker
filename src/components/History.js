import React from "react";
import PropTypes from 'prop-types';

import { BsX } from 'react-icons/bs'

export const History = ({ transactions, setTransactions } ) => {
  const handleDelete = ( id ) => {
    setTransactions(  transactions.filter( t => t.id !== id ) ) 
  }

  return (
    <section className = 'w-100 mb-4'>
      <h2 className = 'fs-6 pb-1 border-bottom border-dark'>Historia</h2>
      {
          transactions.length === 0
          ?
          <p>No se ha registrado transacci√≥nes</p>
          :
          transactions.map( (transaction) => {
            const { desc, amount, id } = transaction;
            const borderColor =  ( amount > 0 ) ? 'border-success': 'border-danger';
            return (
                <div 
                  className = {`${borderColor} border-end  shadow-sm border-2 d-flex align-items-center pt-2 pb-2 ps-2 mb-2 row bg-white container ms-0  me-0 row`}
                  key = { id } 
                >
                  <p className = 'mb-0 col'>
                    { desc }
                  </p>
                  <p className = 'col mb-0 text-end'> {amount} </p>
                  <p
                    className = 'col mb-0 text-end text-danger' 
                  > 
                    <BsX className = 'icon-delete' onClick = { () => handleDelete( id ) }  /> 
                  </p>
                </div>
            )
          } )
      }
    </section>
  );
};


History.propTypes = {
  transactions: PropTypes.array.isRequired,
  setTransactions: PropTypes.func.isRequired
}