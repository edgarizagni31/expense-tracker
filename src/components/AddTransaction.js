import React from 'react';

import { useForm } from '../hooks/useForm';
import { FormItem } from './FormItem';
import Swal from 'sweetalert2';

const createAlert = ( text) => {
  return Swal.fire({
    title: 'Error!',
    text: text,
    icon: 'error',
    confirmButtonText: 'Ok'
  })
}

export const AddTransaction = ({ setTransactions }) => {
  const [transaction, handleInputChange] = useForm({ desc: "", amount: ""});
  const regexNumber = /[1-9]+/;

  const handleSubmitForm = (e) => {
    e.preventDefault();

    if ( transaction.desc.length === 0 && transaction.amount.length === 0) {
      createAlert('Los campos no pueden estar vacios.');
      return;
    }

    if ( transaction.desc === null  || transaction.desc.length === 0) {
      createAlert('Por favor ingrese una descripción.')
      return;
    }

    if ( regexNumber.test(transaction.desc) ) {
      createAlert('Descripción no valida, no puede ser un valor númerico.');
      return;
    }

    if ( transaction.desc.length < 3 && !regexNumber.test(transaction.desc) ) {
      createAlert('Descripción no valida, tiene que tener 3 caracteres como minimo.');
      return;
    }

    if ( transaction.amount === null ) {
      createAlert('Por favor ingrese una cantidad.')
      return;
    } 
    
    if ( transaction.amount.length === 0 || !regexNumber.test(transaction.amount)  ) {
      createAlert('Cantidad no valida, tiene que ser un valor númerico.')
      return;
    }

    setTransactions((t) => [transaction, ...t]);
    document.querySelector("form").reset();
  };

  return (
    <form onSubmit={handleSubmitForm}>
      <h2 className = 'fs-6 pb-1 border-bottom border-dark'>Añadir una nueva transacción</h2>
      <FormItem
        id="desc"
        placeholder="Ingrese una descripción"
        label="Descripción"
        handleInputChange={handleInputChange}
      />
      <FormItem
        id="amount"
        placeholder="Ingrese la cantidad"
        label="Cantidad"
        formText = "negativo(-) gasto, positivo(+) ingreso"
        handleInputChange={handleInputChange}
      />

      <button
        type="submit"
        className="w-100 btn btn-primary mb-3 text-align-center mx-auto"
      >
        Añadir transacción
      </button>
    </form>
  );
};
