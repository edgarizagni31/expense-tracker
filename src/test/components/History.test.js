import { shallow } from 'enzyme'
import { History } from '../../components/History';


describe('test to <History />', () => {
    const setTransaction = jest.fn();

    test('should show correctly', () => {
        const transactions = [];
        const wrapper = shallow(<History  transactions = { transactions } setTransactions = {setTransaction}/>);
        expect(wrapper).toMatchSnapshot();
    })

    test('should show error message when transactions are empty', () => {
        const transactions = [];
        const wrapper = shallow(<History  transactions = { transactions } setTransactions = {setTransaction}/>);
        const errorTag = wrapper.find('p').at(0);

        expect( errorTag.text() ).toBe('No se ha registrado transacciÃ³nes');
    })
    
    test('should show transactions', () => {
        const transactions = [
            {
                desc: 'sueldo',
                amount: '1200',
                id: parseInt( Math.random() * 100 )
            },
            {
                desc: 'comida',
                amount: '-200',
                id: parseInt( Math.random() * 100 )
            }
        ];
        const wrapper = shallow(<History  transactions = { transactions } setTransactions = {setTransaction}/>);
        const divTransanctions = wrapper.find('.border-end');

        expect(divTransanctions.length).toBe(transactions.length);
    })
})
