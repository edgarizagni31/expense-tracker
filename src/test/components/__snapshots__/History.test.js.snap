// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test to <History /> should show correctly 1`] = `
<section
  className="w-100 mb-4"
>
  <h2
    className="fs-6 pb-1 border-bottom border-dark"
  >
    Historia
  </h2>
  <p>
    No se ha registrado transacci√≥nes
  </p>
</section>
`;
